### ISSUE TITLE
Runtime Backpressure

### ISSUE THREAD

**Issue by Theodus** [microsoft/verona#33]
Verona currently does not have a system to prevent a cown's message queue from growing out of memory. I nearly have a working example of backpressure for many cowns sending to a single cown on my fork, which is based on the backpressure system in the Pony runtime. Before I continue to extend this to multimessages in general, I would like to know if this is a feature that the team working on Verona would like to see implemented.

**Comment by sylvanc** [comment]
@Theodus we would definitely like to see this integrated into Verona.

**Comment by SeanTAllen** [comment]
Awesome. This makes me very happy.

**Comment by mjp41** [comment]
So there is an awkward case where the muting used in the Pony back pressure algorithm can lead to a deadlock, if a `when` is part way through acquiring cowns.  I have an internal discussion on this that I will summarise and post here.  There is a solution as well. It may take me a few hours to find and summarise the details.

**Comment by mjp41** [comment]
@Theodus, this is making big assumptions about what you were thinking of implementing, but based on my understanding of the Pony back pressure algorithm this is what goes wrong when you try to extend it to Verona. 

So the core problem is that the Pony back pressure algorithm builds a graph of who muted whom.  This graph is completely unrelated to the order in which cowns are acquired for multi-messages. I think an example something like will highlight this:

```
when (b) { when (a) { ... }}
when (a, b) { ... }
```

So assuming the runtime has chosen `a` to be acquired before `b` in the multi-message order, and that `a` is in a back pressure state, then the first line will cause `b` to sleep, and then the second line will cause `b` to require `a` to be awake before `a` can leave the back pressure state.  This means we have a cycle, and thus deadlock.

**Comment by mjp41** [comment]
Now my notes on a solution to this are below.  It is over a year old, so need to think a bit to see if this would work:

There are two states abnormal states:

* under pressure (message queue too long)
* muted (sent something transitively to an actor that is under pressure)

When you detect back pressure, i.e. you fail to process all message queue in one go:

* alloc new unscheduled cown (the "library", where cowns go to be quiet)
* set state to back pressure, and any one sending here should use the "library"

When no longer under pressure,

* remove under pressure state, and schedule "library"

When you send a message to an actor that is "under pressure" or "muted", when you complete this behaviour,

* set the state of all running cowns to "muted",
* Acquire an epoch (ensures "library" can't be collected) 
* queue a "make noise" message on the "library"
* Release epoch
* do not reschedule these cowns, that is the responsibility of the "library" cown. 

"Make noise" message
* unset "muted" state, 
* lifo schedule muted cown

There is a corner case to think about: what happens if you are "under pressure" and you send a message to a muted cown, do you wait on its "library", or ignore it?  I think you have to wake up your library, and wait on its library.  This would prevent building a cycle.

**Comment by Theodus** [comment]
> the first line will cause `b` to sleep, and then the second line will cause `b` to require `a` to be awake before a can leave the back pressure state.
Just making sure I understand the example. Do you mean that the second line will cause `a` to require `b` to be awake, since `b` has been muted as a result of the first line?

**Comment by mjp41** [comment]
That is what I think I meant a year ago. :-) 

I may have misunderstood the back pressure algorithm you are planning to implement.

**Comment by Theodus** [comment]
I'm running into an issue where, during the cowngc1/4 tests, the "library" cown is collected while it is running its last message because it does a `Cown::release(alloc, this)` in the message body. My questions are: Is it assumed that a cown never releases itself? If so, can the decref be deferred in a way that prevents the collection of the cown during the action?

I have an alternative solution to cleaning up these cowns, but I wanted to ask these questions before I go down that route.

**Comment by mjp41** [comment]
So a thread that has a cown scheduled on it, should own a reference count to it. Send does increfs (or takes a reference count from the caller), to guarantee this invariant.

When a cown becomes deschedule it is decrefed.

Thus, once you send a cown a message, the runtime should deal with ensuring the message arrives before deallocation, and manages those reference counts.

Does that help?

**Comment by Theodus** [comment]
I think so, since that confirms that I'm probably mismanaging the refcounts to the "library cown". Thanks!

**Comment by Theodus** [comment]
So there's a bigger issue causing my confusion. I have been under the assumption that muted cowns should be kept alive while in the "library" cown. However, the cowns are being collected while they are in the "library". Should muted cowns be kept alive like this? And if so, what needs to be done to keep them alive other than `Cown::acquire`?

**Comment by mjp41** [comment]
Hi @theodus, sorry for being slow. Is the collection occuring due to a zero RC or is the leak detector collecting it?  The leak detector requires being called, so if you don't know it is probably the zero RC.

I don't have much time to look into this today.  If you can give me a command line with systematic testing and commit id from you personal fork where this is happening.  I would like to look at the trace tomorrow.

**Comment by Theodus** [comment]
Sure the commit is 63f12a84570b97f65757f3bbb069e65c2781b99b (https://github.com/Theodus/verona/tree/backpressure). The command line is 
```
cmake .. -GNinja -DUSE_ASAN=OFF -DCMAKE_BUILD_TYPE=Debug
ninja
func-sys-cowngc1 --cores 2 --seed 3 --forward 10 --ring 0 --allow_leaks
```
Leaks are allowed because the "library" cowns (currently called `Muting` in the code) are not being collected while this is being investigated.
